[lake] Iceberg lake table support non-string partition keys

### Search before asking

- [x] I searched in the [issues](https://github.com/apache/fluss/issues) and found nothing similar.


### Fluss version

main (development)

### Please describe the bug 🐞

Like #1815 , iceberg lake also should support non-string partition keys.

```
CREATE TABLE fluss_t1 (
    a DATE,
    b STRING,
    PRIMARY KEY (`a`,b) NOT ENFORCED
) PARTITIONED BY (a)
WITH (
  'bucket.num' = '1',
  'table.datalake.enabled' = 'true',
  'table.datalake.freshness' = '1s'
);

INSERT INTO fluss_t1 VALUES (DATE '2025-10-17', 'XXA');
```

Iceberg lake tiering job failed.
```
Caused by: java.lang.ClassCastException: Cannot cast java.lang.String to java.lang.Integer
	at java.lang.Class.cast(Unknown Source) ~[?:?]
	at org.apache.iceberg.StructTransform.get(StructTransform.java:89) ~[fluss-lake-iceberg-0.8-SNAPSHOT.jar:0.8-SNAPSHOT]
	at org.apache.iceberg.PartitionKey.get(PartitionKey.java:30) ~[fluss-lake-iceberg-0.8-SNAPSHOT.jar:0.8-SNAPSHOT]
	at org.apache.iceberg.PartitionSpec.get(PartitionSpec.java:195) ~[fluss-lake-iceberg-0.8-SNAPSHOT.jar:0.8-SNAPSHOT]
	at org.apache.iceberg.PartitionSpec.partitionToPath(PartitionSpec.java:213) ~[fluss-lake-iceberg-0.8-SNAPSHOT.jar:0.8-SNAPSHOT]
	at org.apache.iceberg.LocationProviders$DefaultLocationProvider.newDataLocation(LocationProviders.java:119) ~[fluss-lake-iceberg-0.8-SNAPSHOT.jar:0.8-SNAPSHOT]
	at org.apache.iceberg.io.OutputFileFactory.newOutputFile(OutputFileFactory.java:118) ~[fluss-lake-iceberg-0.8-SNAPSHOT.jar:0.8-SNAPSHOT]
	at org.apache.iceberg.io.OutputFileFactory.newOutputFile(OutputFileFactory.java:113) ~[fluss-lake-iceberg-0.8-SNAPSHOT.jar:0.8-SNAPSHOT]
	at org.apache.iceberg.io.BaseTaskWriter$BaseRollingWriter.openCurrent(BaseTaskWriter.java:332) ~[fluss-lake-iceberg-0.8-SNAPSHOT.jar:0.8-SNAPSHOT]
	at org.apache.iceberg.io.BaseTaskWriter$BaseRollingWriter.<init>(BaseTaskWriter.java:296) ~[fluss-lake-iceberg-0.8-SNAPSHOT.jar:0.8-SNAPSHOT]
	at org.apache.iceberg.io.BaseTaskWriter$RollingFileWriter.<init>(BaseTaskWriter.java:378) ~[fluss-lake-iceberg-0.8-SNAPSHOT.jar:0.8-SNAPSHOT]
	at org.apache.iceberg.io.BaseTaskWriter$BaseEqualityDeltaWriter.<init>(BaseTaskWriter.java:134) ~[fluss-lake-iceberg-0.8-SNAPSHOT.jar:0.8-SNAPSHOT]
	at org.apache.iceberg.io.BaseTaskWriter$BaseEqualityDeltaWriter.<init>(BaseTaskWriter.java:121) ~[fluss-lake-iceberg-0.8-SNAPSHOT.jar:0.8-SNAPSHOT]
	at org.apache.fluss.lake.iceberg.tiering.writer.GenericRecordDeltaWriter$GenericEqualityDeltaWriter.<init>(GenericRecordDeltaWriter.java:76) ~[fluss-lake-iceberg-0.8-SNAPSHOT.jar:0.8-SNAPSHOT]
	at org.apache.fluss.lake.iceberg.tiering.writer.GenericRecordDeltaWriter.<init>(GenericRecordDeltaWriter.java:55) ~[fluss-lake-iceberg-0.8-SNAPSHOT.jar:0.8-SNAPSHOT]
	at org.apache.fluss.lake.iceberg.tiering.writer.TaskWriterFactory.createTaskWriter(TaskWriterFactory.java:89) ~[fluss-lake-iceberg-0.8-SNAPSHOT.jar:0.8-SNAPSHOT]
	at org.apache.fluss.lake.iceberg.tiering.IcebergLakeWriter.createRecordWriter(IcebergLakeWriter.java:90) ~[fluss-lake-iceberg-0.8-SNAPSHOT.jar:0.8-SNAPSHOT]
	at org.apache.fluss.lake.iceberg.tiering.IcebergLakeWriter.<init>(IcebergLakeWriter.java:73) ~[fluss-lake-iceberg-0.8-SNAPSHOT.jar:0.8-SNAPSHOT]
	at org.apache.fluss.lake.iceberg.tiering.IcebergLakeTieringFactory.createLakeWriter(IcebergLakeTieringFactory.java:45) ~[fluss-lake-iceberg-0.8-SNAPSHOT.jar:0.8-SNAPSHOT]
	at org.apache.fluss.flink.tiering.source.TieringSplitReader.getOrCreateLakeWriter(TieringSplitReader.java:313) ~[fluss-flink-1.20-0.8-SNAPSHOT.jar:0.8-SNAPSHOT]
	at org.apache.fluss.flink.tiering.source.TieringSplitReader.forSnapshotSplitRecords(TieringSplitReader.java:391) ~[fluss-flink-1.20-0.8-SNAPSHOT.jar:0.8-SNAPSHOT]
	at org.apache.fluss.flink.tiering.source.TieringSplitReader.fetch(TieringSplitReader.java:133) ~[fluss-flink-1.20-0.8-SNAPSHOT.jar:0.8-SNAPSHOT]
	at org.apache.flink.connector.base.source.reader.fetcher.FetchTask.run(FetchTask.java:58) ~[flink-connector-files-1.20.0.jar:1.20.0]
	at org.apache.flink.connector.base.source.reader.fetcher.SplitFetcher.runOnce(SplitFetcher.java:165) ~[flink-connector-files-1.20.0.jar:1.20.0]
	at org.apache.flink.connector.base.source.reader.fetcher.SplitFetcher.run(SplitFetcher.java:117) ~[flink-connector-files-1.20.0.jar:1.20.0]
	at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) ~[?:?]
	at java.util.concurrent.FutureTask.run(Unknown Source) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) ~[?:?]
	... 1 more
```

```
{
    "format-version": 2,
    "table-uuid": "38f908ff-43e9-46a1-a2e8-c0a38e10b05f",
    "location": "hdfs://namenode:8020/fluss-lake/fluss/fluss_t1",
    "last-sequence-number": 0,
    "last-updated-ms": 1760667212686,
    "last-column-id": 5,
    "current-schema-id": 0,
    "schemas": [
        {
            "type": "struct",
            "schema-id": 0,
            "identifier-field-ids": [
                1,
                2
            ],
            "fields": [
                {
                    "id": 1,
                    "name": "a",
                    "required": true,
                    "type": "date"
                },
                {
                    "id": 2,
                    "name": "b",
                    "required": true,
                    "type": "string"
                },
                {
                    "id": 3,
                    "name": "__bucket",
                    "required": true,
                    "type": "int"
                },
                {
                    "id": 4,
                    "name": "__offset",
                    "required": true,
                    "type": "long"
                },
                {
                    "id": 5,
                    "name": "__timestamp",
                    "required": true,
                    "type": "timestamptz"
                }
            ]
        }
    ],
    "default-spec-id": 0,
    "partition-specs": [
        {
            "spec-id": 0,
            "fields": [
                {
                    "name": "a",
                    "transform": "identity",
                    "source-id": 1,
                    "field-id": 1000
                },
                {
                    "name": "b_bucket",
                    "transform": "bucket[1]",
                    "source-id": 2,
                    "field-id": 1001
                }
            ]
        }
    ],
    "last-partition-id": 1001,
    "default-sort-order-id": 1,
    "sort-orders": [
        {
            "order-id": 1,
            "fields": [
                {
                    "transform": "identity",
                    "source-id": 4,
                    "direction": "asc",
                    "null-order": "nulls-first"
                }
            ]
        }
    ],
    "properties": {
        "write.merge.mode": "merge-on-read",
        "fluss.table.replication.factor": "1",
        "fluss.table.datalake.enabled": "true",
        "fluss.bucket.num": "1",
        "fluss.table.datalake.format": "iceberg",
        "write.delete.mode": "merge-on-read",
        "write.parquet.compression-codec": "zstd",
        "write.update.mode": "merge-on-read",
        "fluss.table.datalake.freshness": "1s"
    },
    "current-snapshot-id": -1,
    "refs": {},
    "snapshots": [],
    "statistics": [],
    "partition-statistics": [],
    "snapshot-log": [],
    "metadata-log": []
}
```

### Solution

_No response_

### Are you willing to submit a PR?

- [x] I'm willing to submit a PR!

--- a/fluss-lake/fluss-lake-iceberg/src/main/java/org/apache/fluss/lake/iceberg/IcebergLakeCatalog.java
+++ b/fluss-lake/fluss-lake-iceberg/src/main/java/org/apache/fluss/lake/iceberg/IcebergLakeCatalog.java
@@ -220,8 +220,7 @@ public class IcebergLakeCatalog implements LakeCatalog {
                 // TODO: Support other types of partition keys
                 throw new InvalidTableException(
                         String.format(
-                                "Iceberg partition key only support string type, "
-                                        + "%s is not string type.",
+                                "Partition key only support string type for iceberg currently. Column `%s` is not string type.",
                                 partitionKey));
             }
             builder.identity(partitionKey);

--- a/fluss-lake/fluss-lake-iceberg/src/test/java/org/apache/fluss/lake/iceberg/IcebergLakeCatalogTest.java
+++ b/fluss-lake/fluss-lake-iceberg/src/test/java/org/apache/fluss/lake/iceberg/IcebergLakeCatalogTest.java
@@ -435,6 +435,6 @@ class IcebergLakeCatalogTest {
                         () -> flussIcebergCatalog.createTable(t1, tableDescriptor.build()))
                 .isInstanceOf(InvalidTableException.class)
                 .hasMessage(
-                        "Iceberg partition key only support string type, c1 is not string type.");
+                        "Partition key only support string type for iceberg currently. Column `c1` is not string type.");
     }
 }

